src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp

```diff
@@ -0,0 +1,310 @@
+
+Module: Dynamic frame condition checking
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+Date: March 2023
+
+\*******************************************************************/
+
+#include "dfcc_root_object.h"
+
+#include <util/byte_operators.h>
+#include <util/cprover_prefix.h>
+#include <util/expr.h>
+#include <util/expr_util.h>
+#include <util/message.h>
+#include <util/namespace.h>
+#include <util/pointer_expr.h>
+#include <util/std_code.h>
+
+#include <langapi/language_util.h>
+
+/// Translates object slice expressions found in assigns clause targets to
+/// dereference expressions, so that object_descriptor_exprt can be used to
+/// compute the base object expression.
+static const exprt slice_op_to_deref(const exprt &expr)
+{
+  if(can_cast_expr<side_effect_expr_function_callt>(expr))
+  {
+    auto function_call_expr = to_side_effect_expr_function_call(expr);
+    auto function_expr = function_call_expr.function();
+    INVARIANT(
+      function_expr.id() == ID_symbol,
+      "no function pointer calls loop assigns clause targets");
+    auto function_id = to_symbol_expr(function_expr).get_identifier();
+    INVARIANT(
+      function_id == CPROVER_PREFIX "assignable" ||
+        function_id == CPROVER_PREFIX "object_whole" ||
+        function_id == CPROVER_PREFIX "object_from" ||
+        function_id == CPROVER_PREFIX "object_upto",
+      "only built-in function calls in loop assigns clause targets");
+    return dereference_exprt(
+      skip_typecast(function_call_expr.arguments().at(0)));
+  }
+  else
+  {
+    return expr;
+  }
+}
+
+const exprt
+dfcc_loop_assigns_target_root_object(const exprt &expr, const namespacet &ns)
+{
+  exprt fixed = slice_op_to_deref(expr);
+  object_descriptor_exprt ode;
+  ode.build(expr, ns);
+  return ode.root_object();
+}
+
+/// Returns `arr` if \p expr matches `arr[idx]` where `arr` has an array type,
+/// `nullopt` otherwise.
+static optionalt<exprt> match_array_index(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_index)
+  {
+    const exprt &op0 = skipped.operands().at(0);
+    if(op0.type().id() == ID_array)
+    {
+      return op0;
+    }
+  }
+  return nullopt;
+}
+
+/// Returns `arr` if \p expr matches `&arr[idx]` where `arr` has an array type,
+/// `nullopt` otherwise.
+static optionalt<exprt> match_address_of_array_index(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_address_of)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
+    return match_array_index(subexpr);
+  }
+  return nullopt;
+}
+
+/// Returns `arr` if \p expr matches `*(&arr[idx])` where `arr` has an array
+/// type, `nullopt` otherwise.
+static optionalt<exprt> match_deref_of_address_of_array_index(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_dereference)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
```

Comment:
As above: please avoid `.operands.at(0)` and use proper expression APIs instead.

Reply:
fixed